### PR TITLE
Part of JARVIS-705: Add assistant timezone resolution primitives

### PR DESCRIPTION
**Pull request by awlevin**
## Summary
- Adds canonical timezone validation for assistant temporal config.
- Introduces reusable turn timezone resolution for later client timezone work.
- Covers precedence and compatibility behavior with focused tests.

Part of JARVIS-705.
Part of plan: assistant-local-timezone.md (PR 1 of 8)

---

<a href="https://app.devin.ai/review/vellum-ai/vellum-assistant/pull/29670" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by awlevin** [comment]
<details>
<summary>Plan papertrail: Assistant Local Timezone Grounding (JARVIS-705)</summary>

# Assistant Local Timezone Grounding (JARVIS-705)

## Overview
This plan fixes assistant turns that sometimes reason in UTC by making client/device timezone a first-class runtime signal while preserving `ui.userTimezone` as the explicit manual override. It covers the assistant runtime, HTTP message contract, natural-language timezone persistence via the existing config CLI, macOS client behavior, tests, docs, and release notes.

Planning note: the checked-out assistant repo was `main...origin/main [behind 2]` with unrelated local changes in `credential-executor/src/cli.ts`, `meta/bun.lock`, `Untitled`, and `clients/ios/`; the sibling platform repo was `main...origin/main [behind 12]`. I did not pull either repo because this is a local planning artifact and the assistant worktree is dirty.

## PR 1: Add Timezone Resolution Primitives
### Depends on
None

### Branch
jarvis-705-utc-time/pr-1-timezone-resolution

### Title
Part of JARVIS-705: Add assistant timezone resolution primitives

### Files
- assistant/src/config/schemas/platform.ts
- assistant/src/daemon/date-context.ts
- assistant/src/__tests__/date-context.test.ts

### Implementation steps
1. In `assistant/src/config/schemas/platform.ts`, extend `UiConfigSchema` with `detectedTimezone?: string`. Keep the existing `ui.userTimezone` field backwards-compatible by allowing `undefined` and the existing empty-string clearing sentinel.
2. Add a local timezone validator/canonicalizer used by both `userTimezone` and `detectedTimezone`. It should accept valid IANA identifiers and `""`, and reject non-empty invalid strings with a clear schema error.
3. Update the schema descriptions so `ui.userTimezone` is documented as the manual timezone override used for assistant temporal grounding, not just display formatting.
4. In `assistant/src/daemon/date-context.ts`, export a reusable timezone canonicalization helper instead of keeping all validation private to `formatTurnTimestamp`.
5. Add `clientTimezone?: string | null` and `detectedTimezone?: string | null` to `TemporalContextOptions`.
6. Add a `resolveTurnTimezoneContext(options)` helper returning canonical values for `configuredUserTimezone`, `clientTimezone`, `detectedTimezone`, `hostTimezone`, `effectiveTimezone`, and the selected source.
7. Update `formatTurnTimestamp()` to use `resolveTurnTimezoneContext()` while preserving the current string format: `YYYY-MM-DD (Weekday) HH:MM:SS ±HH:MM (IANA/Zone)`.
8. Add tests proving normal turn precedence is `ui.userTimezone`, client timezone, `ui.detectedTimezone`, host timezone, then UTC fallback.
9. Do not reintroduce dormant recalled/profile timezone precedence into the normal turn path. The bug this plan targets is current-turn grounding, so per-message client timezone and persisted detected timezone should be the automatic sources.
10. Add tests proving empty strings are ignored for runtime resolution but accepted for config compatibility.
11. If `formatTurnTimestamp({ timeZone })` remains as a low-level test/internal option, do not document it as a user-facing or turn-level precedence layer.

### Acceptance criteria
- `bun test assistant/src/__tests__/date-context.test.ts` passes.
- Existing configs that use `ui.userTimezone: ""` to clear the setting still parse and still produce no configured timezone.
- Invalid non-empty timezone strings are rejected by the config schema for both `ui.userTimezone` and `ui.detectedTimezone`.
- `formatTurnTimestamp()` output remains backwards-compatible for existing callers.

## PR 2: Store Client Timezone on Conversation Transport
### Depends on
PR 1

### Branch
jarvis-705-utc-time/pr-2-transport-timezone

### Title
Part of JARVIS-705: Carry client timezone through conversation transport

### Files
- assistant/src/daemon/message-types/conversations.ts
- assistant/src/daemon/conversation.ts
- assistant/src/daemon/conversation-process.ts

### Implementation steps
1. Add optional `clientTimezone?: string` to `BaseTransportMetadata` in `assistant/src/daemon/message-types/conversations.ts`.
2. Add `clientTimezone?: string` to the `Conversation` class state in `assistant/src/daemon/conversation.ts`.
3. Add `clientTimezone?: string` and `applyClientTimezoneFromTransport(transport: ConversationTransportMetadata): void` to the `ProcessConversationContext` interface in `assistant/src/daemon/conversation-process.ts`.
4. Implement `Conversation.applyClientTimezoneFromTransport()` in `assistant/src/daemon/conversation.ts`. It should canonicalize with the helper from PR 1 and set `this.clientTimezone` only when the transport value is valid; invalid or absent values should clear the previous per-turn value so reused conversations do not leak stale device timezone.
5. In `assistant/src/daemon/conversation-process.ts`, call `conversation.applyClientTimezoneFromTransport(next.transport)` next to the existing `setTransportHints()` and `applyHostEnvFromTransport()` calls for single-message drains.
6. In the batched drain path, call the same method using the head message transport, matching the existing behavior for transport hints and host env.

### Acceptance criteria
- Single-message and batched turns apply the current message transport timezone before `runAgentLoop()`.
- A conversation reused across web, macOS, and channel ingress does not retain an earlier client timezone when the current transport omits it.
- TypeScript type-checks with `.js` import extensions preserved.

## PR 3: Accept Client Timezone on HTTP Messages
### Depends on
PR 2

### Branch
jarvis-705-utc-time/pr-3-message-client-timezone

### Title
Part of JARVIS-705: Accept client timezone on message posts

### Files
- assistant/src/runtime/routes/conversation-routes.ts
- assistant/src/__tests__/http-user-message-parity.test.ts

### Implementation steps
1. In the `/v1/messages` body shape inside `assistant/src/runtime/routes/conversation-routes.ts`, add optional `clientTimezone?: string`.
2. When constructing `transport`, pass `clientTimezone` through for both host-proxy and non-host-proxy metadata variants.
3. Canonicalize `clientTimezone` before putting it on transport. Do not reject an otherwise valid message solely because this optional timezone is invalid; drop invalid values instead and let the host/config fallback handle the turn.
4. Add HTTP route tests proving a valid `clientTimezone` reaches `getOrCreateConversation(..., { transport })`.
5. Add a regression test proving invalid `clientTimezone` does not 400 the message request and does not reach transport.

### Acceptance criteria
- `bun test assistant/src/__tests__/http-user-message-parity.test.ts` passes.
- Existing clients that do not send `clientTimezone` continue to work unchanged.
- Invalid optional timezone input cannot poison later turns.

## PR 4: Inject Timezone Mismatch Context
### Depends on
PR 1, PR 2

### Branch
jarvis-705-utc-time/pr-4-turn-context-timezone

### Title
Part of JARVIS-705: Inject effective and device timezone context

### Files
- assistant/src/daemon/conversation-agent-loop.ts
- assistant/src/daemon/conversation-runtime-assembly.ts
- assistant/src/__tests__/conversation-runtime-assembly.test.ts
- assistant/src/__tests__/conversation-agent-loop.test.ts

### Implementation steps
1. In `assistant/src/daemon/conversation-agent-loop.ts`, read `getConfig().ui.userTimezone`, `getConfig().ui.detectedTimezone`, `ctx.clientTimezone`, and the host timezone.
2. Use `resolveTurnTimezoneContext()` from PR 1 to build the timestamp and retain the canonical configured/client/detected values.
3. Extend `UnifiedTurnContextOptions` in `assistant/src/daemon/conversation-runtime-assembly.ts` with optional timezone context fields, for example `configuredUserTimezone?: string | null` and `clientTimezone?: string | null`.
4. In `buildUnifiedTurnContextBlock()`, emit `configured_user_timezone: <zone>` and `client_device_timezone: <zone>` only when an explicit `ui.userTimezone` exists and the current client/detected timezone differs from it.
5. Keep `current_time:` in the effective timezone. If `ui.userTimezone` is set, it should remain the effective timezone even when the device timezone differs.
6. Add a compact persistence affordance line only in the mismatch case, for example `timezone_update_available: after explicit user confirmation, persist client_device_timezone with \`assistant config set ui.userTimezone "<zone>"\``. This lets the assistant offer to fix the setting through natural language without adding a new registered tool.
7. Keep mismatch context factual and compact. Do not add broad behavioral instructions to the system prompt.
8. Add tests for no extra lines when there is no manual timezone override, no extra lines when configured and client timezone match, both timezone lines when they differ, and the CLI affordance line only in the mismatch case.
9. Add agent-loop tests proving the runtime passes `ctx.clientTimezone` and `ui.detectedTimezone` into the turn context calculation.

### Acceptance criteria
- `bun test assistant/src/__tests__/conversation-runtime-assembly.test.ts --grep timezone` passes.
- `bun test assistant/src/__tests__/conversation-agent-loop.test.ts --grep timezone` passes.
- `<turn_context>` remains compact and only adds mismatch context when it is useful.
- Explicit user timezone always wins over device timezone for `current_time`.
- The assistant can ask the user before changing timezone and then persist the update by running `assistant config set ui.userTimezone "<client_device_timezone>"`.

## PR 5: Send macOS Device Timezone Per Message
### Depends on
PR 3

### Branch
jarvis-705-utc-time/pr-5-macos-message-timezone

### Title
Part of JARVIS-705: Send macOS client timezone with messages

### Files
- clients/shared/Network/MessageClient.swift
- clients/macos/vellum-assistantTests/MessageClientTimezoneTests.swift

### Implementation steps
1. In `clients/shared/Network/MessageClient.swift`, add a small internal helper that returns `TimeZone.autoupdatingCurrent.identifier` when it is non-empty.
2. Add `clientTimezone` to the HTTP message body built by `MessageClient.sendMessage(...)`.
3. Extract message body construction into an internal static helper if necessary so the timezone behavior can be tested without performing network I/O.
4. Add `MessageClientTimezoneTests.swift` proving the built body includes the local timezone and keeps existing fields such as `sourceChannel`, `interface`, `hostHomeDir`, and `hostUsername`.
5. Avoid adding timezone to attachment uploads or unrelated client requests.

### Acceptance criteria
- Targeted Swift tests for the new message body helper pass.
- macOS HTTP chat sends carry a valid IANA `clientTimezone`.
- No user-facing macOS text changes are introduced in this PR.

## PR 6: Add macOS Automatic Timezone Settings
### Depends on
PR 1

### Branch
jarvis-705-utc-time/pr-6-macos-timezone-settings

### Title
Part of JARVIS-705: Add automatic timezone mode to macOS settings

### Files
- clients/macos/vellum-assistant/Features/Settings/SettingsStore.swift
- clients/macos/vellum-assistant/Features/Settings/SettingsAppearanceTab.swift
- clients/macos/vellum-assistantTests/SettingsStoreUserTimezoneTests.swift

### Implementation steps
1. In `SettingsStore.swift`, add `@Published var detectedTimezone: String?` loaded from `ui.detectedTimezone`.
2. Add a `saveDetectedTimezone(_:)` method that canonicalizes an IANA identifier and PATCHes `{"ui": {"detectedTimezone": "<zone>"}}`.
3. On config load and reconnect, hydrate both `userTimezone` and `detectedTimezone`.
4. On startup or settings appearance, save `TimeZone.autoupdatingCurrent.identifier` to `ui.detectedTimezone` when it differs from the stored detected value.
5. In `SettingsAppearanceTab.swift`, change the timezone card to show an Automatic/manual control. Automatic should clear `ui.userTimezone` and display the detected device timezone. Manual should preserve the existing searchable closest-city picker and write `ui.userTimezone`.
6. Keep existing manual timezone semantics: choosing a city writes `ui.userTimezone`; clearing manual returns to automatic behavior.
7. Update `SettingsStoreUserTimezoneTests.swift` to cover loading detected timezone, saving detected timezone, automatic-mode clearing of manual override, and invalid detected timezone handling.

### Acceptance criteria
- Existing users with `ui.userTimezone` see manual mode selected and keep the same effective timezone.
- Users without `ui.userTimezone` get automatic mode using the current device timezone.
- Clearing manual mode does not delete unrelated `ui` settings.
- Targeted macOS settings tests pass.

## PR 7: Update Assistant Timezone Docs
### Depends on
PR 4, PR 5, PR 6

### Branch
jarvis-705-utc-time/pr-7-timezone-docs

### Title
Part of JARVIS-705: Document assistant timezone precedence

### Files
- assistant/docs/architecture/memory.md
- clients/ARCHITECTURE.md

### Implementation steps
1. In `assistant/docs/architecture/memory.md`, update the `<turn_context>` timezone section to document precedence: manual `ui.userTimezone`, current client timezone, detected timezone, host timezone.
2. Document the optional mismatch lines emitted when explicit user timezone differs from the current client/detected timezone.
3. In `clients/ARCHITECTURE.md`, update the Settings Persistence example to include `ui.detectedTimezone`.
4. Update the prose so `ui.userTimezone` is clearly the manual override and `ui.detectedTimezone` is the automatic device-derived default.
5. Document that `ui.userTimezone` is a historical namespace for a runtime-affecting setting, not a purely presentational UI preference. It is authoritative for `current_time` when set.
6. Document the natural-language update path: after explicit confirmation, the assistant can persist a timezone correction by running `assistant config set ui.userTimezone "<IANA zone>"`.

### Acceptance criteria
- Docs match the implemented field names and turn context behavior.
- No architecture docs claim web localStorage is the source of truth for assistant timezone.

## PR 8: Add Release Notes Migration
### Depends on
PR 7

### Branch
jarvis-705-utc-time/pr-8-timezone-release-note

### Title
Closes JARVIS-705: Add local timezone release note

### Files
- assistant/src/workspace/migrations/067-release-notes-local-timezone.ts
- assistant/src/workspace/migrations/registry.ts
- assistant/src/__tests__/workspace-migration-067-release-notes-local-timezone.test.ts

### Implementation steps
1. Add the next available release-note workspace migration. As of this plan, the next sequence is `067`, but verify before implementation and increment if another migration lands first.
2. The migration should append a short note to `UPDATES.md` explaining that the assistant now defaults to the device timezone and respects manual timezone overrides across clients.
3. Include an HTML marker such as `<!-- release-note-id:067-release-notes-local-timezone -->` and short-circuit if the marker is already present.
4. Register the migration at the end of `WORKSPACE_MIGRATIONS` in `registry.ts` without reordering existing entries.
5. Add a focused migration test proving the note appends once and does not duplicate when re-run.

### Acceptance criteria
- `bun test assistant/src/__tests__/workspace-migration-067-release-notes-local-timezone.test.ts` passes.
- The migration is idempotent via both the runner and the in-file marker.
- The final PR body uses `Closes JARVIS-705`.

</details>

File: assistant/src/config/schemas/platform.ts [hunk 1]
```diff
@@ -1,5 +1,38 @@
...
+  if (trimmed.length === 0) {
+    return "";
+  }
+  if (/^[+-]/.test(trimmed) || /^(?:UTC|GMT)[+-]/i.test(trimmed)) {
+    return null;
+  }
+  try {
+    return new Intl.DateTimeFormat("en-US", {
+      timeZone: trimmed,
+    }).resolvedOptions().timeZone;
```
[the remark this answers is not in the document]
Comment (awlevin) [reply assistant/src/config/schemas/platform.ts:17]: Addressed in direct push to jarvis-705-utc-time/pr-1-timezone-resolution

File: assistant/src/config/schemas/platform.ts [hunk 2]
```diff
@@ -56,11 +89,15 @@ export const DaemonConfigSchema = z
 
 export const UiConfigSchema = z
   .object({
-    userTimezone: z
-      .string({ error: "ui.userTimezone must be a string" })
+    userTimezone: timezoneConfigField("ui.userTimezone")
+      .optional()
+      .describe(
+        "Manual IANA timezone override used for assistant temporal grounding and date/time display (e.g. 'America/New_York'). Use an empty string to clear the setting.",
+      ),
```
[the remark this answers is not in the document]
Comment (awlevin) [reply assistant/src/config/schemas/platform.ts:99]: Addressed in direct push to jarvis-705-utc-time/pr-1-timezone-resolution

File: assistant/src/daemon/date-context.ts [hunk 3]
```diff
@@ -289,36 +326,49 @@ function formatLocalDate(date: Date, timeZone: string): string {
...
+
+  return {
+    configuredUserTimezone,
+    clientTimezone,
+    detectedTimezone,
+    hostTimezone,
+    effectiveTimezone: selected?.timeZone ?? "UTC",
+    source: selected?.source ?? "utcFallback",
+  };
+}
```
[the remark this answers is not in the document]
Comment (awlevin) [reply assistant/src/daemon/date-context.ts:357]: Addressed in direct push to jarvis-705-utc-time/pr-1-timezone-resolution

**Comment by awlevin** [comment]
@codex review d0c61be6da3841d3ea8bae553e6952e0bb5996b9
@devin review d0c61be6da3841d3ea8bae553e6952e0bb5996b9

**Comment by awlevin** [comment]
Review gate note: all bot review threads are resolved at thread level after commit d0c61be6da3841d3ea8bae553e6952e0bb5996b9. The remaining changes_requested aggregate is stale from flat inline comments/reactions, so proceeding with merge into the feature branch.
